pydantic_ai_slim/pydantic_ai/prompt_config.py

```diff
@@ -0,0 +1,414 @@
+    Pydantic AI sends to the model. This includes:
+
+    - **Prompt Templates**: Messages for retry prompts, tool return confirmations,
+      validation errors, and other system-generated text via [`PromptTemplates`][pydantic_ai.PromptTemplates].
+    - **Tool Configuration**: Tool descriptions, parameter descriptions, and other
+      tool metadata - allowing you to override descriptions and args for tools at the agent level.
+
+    This allows you to fully customize how your agent communicates with the model
+    without modifying the underlying tool or agent code.
+
+    Example:
+        ```python
+        from pydantic_ai import Agent, PromptConfig, PromptTemplates
+
+        agent = Agent(
+                    'openai:gpt-4o',
+                    prompt_config=PromptConfig(
+                        templates=PromptTemplates(
+                            validation_errors_retry_prompt='Please correct the errors and try again.',
+                            final_result_processed='Result received successfully.',
+                        ),
+                    ),
+                )
+        ```
+
+    Attributes:
+        templates: Templates for customizing system-generated messages like retry prompts,
+            tool return confirmations, and validation error messages.
+        tool_config: Configuration for customizing tool descriptions and metadata.
+    """
+
+    templates: PromptTemplates | None = None
+    """Templates for customizing system-generated messages sent to the model.
+
+    See [`PromptTemplates`][pydantic_ai.PromptTemplates] for available template options.
+    """
+
+    tool_config: dict[str, ToolConfig] | None = None
+    """Configuration for customizing tool descriptions and metadata, keyed by tool name.
+    See [`ToolConfig`][pydantic_ai.ToolConfig] for available configuration options.
+    """
+
+    @staticmethod
+    async def generate_prompt_config_from_agent(agent: Agent[Any, Any], model: Model) -> PromptConfig:
+        """Generate a PromptConfig instance based on an Agent instance.
+
+        This fills in per-tool metadata and default templates, producing a `PromptConfig` that can be
+        used as a starting point for prompt optimizers.
+
+        """
+        tool_config: dict[str, ToolConfig] = {}
+
+        prompt_templates: PromptTemplates = PromptTemplates(
+            final_result_processed=DEFAULT_FINAL_RESULT_PROCESSED,
+            output_tool_not_executed=DEFAULT_OUTPUT_TOOL_NOT_EXECUTED,
+            output_validation_failed=DEFAULT_OUTPUT_VALIDATION_FAILED,
+            function_tool_not_executed=DEFAULT_FUNCTION_TOOL_NOT_EXECUTED,
+            tool_denied=DEFAULT_TOOL_CALL_DENIED,
+            validation_errors_retry_prompt=DEFAULT_MODEL_RETRY,
+            tool_retry_prompt=DEFAULT_MODEL_RETRY,
+            no_tool_retry_prompt=DEFAULT_MODEL_RETRY,
+            prompted_output_template=DEFAULT_PROMPTED_OUTPUT_TEMPLATE,
+            description_template=None,
```

Comment:
Why none?

Reply:
Added a default function to resolve via one of these two based on the content.